samza-core/src/main/java/org/apache/samza/container/RunLoop.java

```diff
@@ -805,6 +877,52 @@ private boolean checkEndOfStream() {
       return processingSspSet.isEmpty();
     }
 
+    private boolean shouldDrain() {
+      if (endOfStream) {
+        return false;
+      }
+
+      if (!pendingEnvelopeQueue.isEmpty()) {
+        PendingEnvelope pendingEnvelope = pendingEnvelopeQueue.peek();
+        IncomingMessageEnvelope envelope = pendingEnvelope.envelope;
+
+        if (envelope.isDrain()) {
+          final DrainMessage message = (DrainMessage) envelope.getMessage();
+          if (!message.getDeploymentId().equals(deploymentId)) {
+            // Removing the drain message from the pending queue as it doesn't match with the current deploymentId
+            // Removing it will ensure that it is not picked up by process()
+            pendingEnvelopeQueue.remove();
+          } else {
+            if (elasticityFactor <= 1) {
+              SystemStreamPartition ssp = envelope.getSystemStreamPartition();
+              processingSspSetToDrain.remove(ssp);
+            } else {
+              // if envelope contains drain message,
+              // the ssp of envelope should be removed from task's processing set irresp of keyBucket
+              SystemStreamPartition sspOfEnvelope = envelope.getSystemStreamPartition();
+              Optional<SystemStreamPartition> ssp = processingSspSetToDrain.stream()
+                  .filter(sspInSet -> sspInSet.getSystemStream().equals(sspOfEnvelope.getSystemStream())
+                      && sspInSet.getPartition().equals(sspOfEnvelope.getPartition()))
+                  .findFirst();
+              ssp.ifPresent(processingSspSetToDrain::remove);
+            }
+
+            if (!hasIntermediateStreams) {
```

Comment:
Is this condition strong enough? Seems to me that we want to check whether this drain message is coming from an input stream v.s. intermediate stream? The task might consume both input stream and intermediate stream partitions.

Reply:
This condition is necessary for high level API otherwise RunLoop will simply stop the task when drain message is seen for the final SSP. We want the drain message to go be processed by the high-level API DAG. This holds true for source SSPs and intermediate SSPs alike.